server: hot ranges api

koorosh opened the pull request:
This change implements second version of hot ranges api that's required
for UI to represent enhanced Hot Ranges view.

It reuses former implementation but doesn't expose sensitive internal data.
Instead, it provides flat structure of fields required for ui.

New `HotRangeV2` endpoint is established to:
- maintain backward compatibility
- do not expose internal data as it's done in `HotRange` endpoint

Release note: None

Release justification: bug fixes and low-risk updates to new functionality

koorosh committed:
server: hot ranges api

This change implements second version of hot ranges api that's required
for UI to represent enhanced Hot Ranges page.

Before, Hot ranges (under Advanced debugging page) used former version of
HotRanges api that provided information about hot ranges with internal/
sensitive data (see: https://github.com/cockroachdb/cockroach/issues/53212)
that should not be exposed to users.

Now, in addition to existing endpoint, additional one is implemented that is
based on a former version and provides hot ranges information that is only
needed for Hot Ranges page (range id, qps, table, db, and index names for
particular range).

The list of hot ranges and their QPS is provided by `HotRanges` service and
then information like DB, table and index names are retrieved from range's
`StartKey` that might include this info or not (in case if it's meta range,
or range that stores index itself for instance).

`HotRange` and `HotRangeV2` services expect the same request type as an
argument but return different responses.
`HotRangeV2` service returns a flat list of hot ranges instead of grouped
ranges per node/store.

Release note: None

server: add leaseholder node id to hot ranges api

Current change extends `statuspb.HotRangesResponse` to
include `LeaseholderNodeID` field to indicate the
node id that contains leaseholder replica for
current hot range.

This change was made in `localHotRanges` function (that is
used by `HotRanges` that in turn used by `HotRangeV2` service)
to reuse existing logic of iteration over the stores and querying
hot ranges. It extends its response by `LeaseholderNodeID` field.
Otherwise, the same logic should be implemented in `HotRangeV2`
service by calling `VisitStores` iterator.

Release note: None

Release justification: bug fixes and low-risk updates
to new functionality

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74377)
<!-- Reviewable:end -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @koorosh)

___
*[pkg/server/serverpb/status.proto, line 1721 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Msk-CeCCRZfotnNGI9Z:-Msk-CeCCRZfotnNGI9_:b9tx7q1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6f43341db514c83f1b4246cfa6c582e6e5dcf51/pkg/server/serverpb/status.proto#L1721)):*
> ```protobuf
>   }
> 
>   rpc HotRanges2(HotRangesRequest) returns (HotRangesResponseV2) {
> ```

Can we reconsider the placement of this API call? We already have an `/api/v2/` server that new stuff should go under and I'd prefer to use that if we can.

See https://github.com/cockroachdb/cockroach/blob/master/pkg/server/api_v2.go


<!-- Sent from Reviewable.io -->

koorosh committed:
server: update hot ranges api/v2 version

Before, `listHotRanges` request handler of `apiV2Server` relied
on `HotRanges` service that is now should be replaced by new
`HotRangesV2` implementation.

Current change reuses HotRangeV2 service in `api/v2/ranges/hot`
api. It allows to share the same logic between REST and gRPC
endpoints and gradually migrate to new version of API.

Release note: None

Release justification: bug fixes and low-risk updates
to new functionality

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dhartunian)

___
*[pkg/server/serverpb/status.proto, line 1721 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Msk-CeCCRZfotnNGI9Z:-Mt2UH_d89tVdXOWC0G5:b5pus2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/f42a85a5abfec2757e6869f8cd050fad3658b377/pkg/server/serverpb/status.proto#L1721)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Can we reconsider the placement of this API call? We already have an `/api/v2/` server that new stuff should go under and I'd prefer to use that if we can.

See https://github.com/cockroachdb/cockroach/blob/master/pkg/server/api_v2.go
</blockquote></details>

Done. I updated `/api/v2/ranges/hot` api to provide updated version of Hot ranges, but I would prefer to keep `/_status/v2/hotranges` as well until `/api/v2/ranges/hot` handles authorization validation (for secure and insecure modes). 


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
Nice work! It's great to see this coming together.

One general question I wanted to pose here is have we considered how this might work in a serverless context? With the upcoming project to improve tenant-level observability, we'd ideally like to be able to generate a hot ranges report for a specific serverless tenant. We're hoping to build some level of debug.zip functionality at the tenant level, and one of the pieces that we think will be useful for such a bundle is a hot ranges report, since this is relevant within the SQL domain.

This is not necessarily in scope for this PR, but we should think about it now instead of later to avoid having to build a `HotRangesV3` endpoint a few months from now :sweat_smile:. What would we need to do to make this work at the tenant-level?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-:-Mt4bRay9AT8OVpxHjQq:b-e71w4a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, and @koorosh)

___
*[pkg/server/status.go, line 2073 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt4_Z7_2J_FGAjWCx1m:-Mt4_Z7_2J_FGAjWCx1n:b-vcjara) ([raw file](https://github.com/cockroachdb/cockroach/blob/896d8a409804e251fd6c7f73255df5f4f2d8cfa5/pkg/server/status.go#L2073)):*
> ```Go
> 	}
> 
> 	dbNames := make(map[uint32]string)
> ```

Can we possibly make a `struct` type to more clearly keep track of all this information? 

It seems like the key for all these maps is the `catalog.Descriptor` ID, so instead of having multiple maps, can we have a single struct type that encompasses all these attributes (db names, table names, index names, and parent ID), and then just use a single map? Something like

```
type HotRangeReportMeta struct {
    dbName string
    tableName string
    indexNames map[uint32]string
    parentId int32
}
rangeReportMetas := make(map[uint32]HotRangeReportMeta)
...
for _, desc := range descrs {
    rangeReportMetas[uint32(desc.GetId())] = HotRangeReportMeta{
        ...
    }
} 
```

Then the maintainer can get all the info they need with a single map access. Just an idea, let me know what you think. We could also potentially define functions on this type to encapsulate some of the more complex access patterns of this data, although maybe that's overkill :D 



_Code quote:_
```Go
	dbNames := make(map[uint32]string)
	tableNames := make(map[uint32]string)
	indexNames := make(map[uint32]map[uint32]string)
	parents := make(map[uint32]uint32)
```
___
*[pkg/server/status.go, line 2106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt4clwCDuSYv3Petrdg:-Mt4clwCDuSYv3Petrdh:b-22e7gz) ([raw file](https://github.com/cockroachdb/cockroach/blob/896d8a409804e251fd6c7f73255df5f4f2d8cfa5/pkg/server/status.go#L2106)):*
> ```Go
> 
> 	var ranges []*serverpb.HotRangesResponseV2_HotRange
> 	// TODO (koorosh): how to flatten triple nested loop?
> ```

Nothing stands out to me that would allow us to avoid this, I don't see it as too big of an issue, personally :) 



_Code quote:_
```Go
// TODO (koorosh): how to flatten triple nested loop?
```


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> Reviewed 1 of 6 files at r1, 8 of 8 files at r4, 5 of 5 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @koorosh)

___
*[-- commits, line 8 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8uguQCZfvIRObtVK2):*
nit: can you reword this a bit? It seems there are two changes being made with `v2`: not exposing sensitive data and creating a flat structure. Please make that a bit more clear.
___
*[-- commits, line 16 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8uytK1rh6htaiTE-I):*
nit: use backticks for fields
___
*[-- commits, line 21 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8v3lX7-LNY3Snw825):*
nit: can you add a note about why this would lead to additional store scans?
___
*[-- commits, line 28 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt90NrWEDqxNYruuGg4):*
nit: can you clarify what the implementation was before and after the change. It's unclear from this sentence.
___
*[docs/generated/swagger/spec.json, line 1779 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8ysLV8zrq501VFDLw:-Mt8ysLW3-yc1t7rhhn1:budegao) ([raw file](https://github.com/cockroachdb/cockroach/blob/76d43e646d32a9ff458620589a4801129dcae9e6/docs/generated/swagger/spec.json#L1779)):*
> ```JSON
>     },
>     "hotRangeInfo": {
>       "description": "Hot range details",
> ```

This should be more detailed.
___
*[pkg/server/status.go, line 2065 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8uL5AAtfMsH7VfCbn:-Mt8uL5AAtfMsH7VfCbo:b-sn5wup) ([raw file](https://github.com/cockroachdb/cockroach/blob/896d8a409804e251fd6c7f73255df5f4f2d8cfa5/pkg/server/status.go#L2065)):*
> ```Go
> }
> 
> func (s *statusServer) HotRanges2(
> ```

nit: docstring
___
*[pkg/server/serverpb/status.proto, line 1721 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Msk-CeCCRZfotnNGI9Z:-Mt8uWKA1VkvwxgSxWpk:b-81nlyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f42a85a5abfec2757e6869f8cd050fad3658b377/pkg/server/serverpb/status.proto#L1721)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

Done. I updated `/api/v2/ranges/hot` api to provide updated version of Hot ranges, but I would prefer to keep `/_status/v2/hotranges` as well until `/api/v2/ranges/hot` handles authorization validation (for secure and insecure modes). 
</blockquote></details>

SGTM 
___
*[pkg/server/serverpb/status.proto, line 1126 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8vJPD-fdhxopMQOqq:-Mt8vJPD-fdhxopMQOqr:b-sn5wup) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fc6c1aabe03af0a466500806c54ee96f15a9c22/pkg/server/serverpb/status.proto#L1126)):*
> ```protobuf
> 
> message HotRangesResponseV2 {
>   message HotRange {
> ```

nit: docstring


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dhartunian)

___
*[-- commits, line 8 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8uguQCZfvIRObtVK2):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: can you reword this a bit? It seems there are two changes being made with `v2`: not exposing sensitive data and creating a flat structure. Please make that a bit more clear.
</blockquote></details>

Done.
___
*[-- commits, line 16 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8uytK1rh6htaiTE-I):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: use backticks for fields
</blockquote></details>

Done.
___
*[-- commits, line 21 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8v3lX7-LNY3Snw825):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: can you add a note about why this would lead to additional store scans?
</blockquote></details>

Done.
___
*[-- commits, line 28 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt90NrWEDqxNYruuGg4):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: can you clarify what the implementation was before and after the change. It's unclear from this sentence.
</blockquote></details>

Done.
___
*[docs/generated/swagger/spec.json, line 1779 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8ysLV8zrq501VFDLw:-MtDXn6AB_k-1TqCXMlG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/76d43e646d32a9ff458620589a4801129dcae9e6/docs/generated/swagger/spec.json#L1779)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This should be more detailed.
</blockquote></details>

Done.
___
*[pkg/server/status.go, line 2065 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt8uL5AAtfMsH7VfCbn:-MtDkFCr7wNuCeeGucUV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/896d8a409804e251fd6c7f73255df5f4f2d8cfa5/pkg/server/status.go#L2065)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: docstring
</blockquote></details>

Done.
___
*[pkg/server/status.go, line 2073 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt4_Z7_2J_FGAjWCx1m:-Mt8uxnt4D7EqI4J9ur9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/896d8a409804e251fd6c7f73255df5f4f2d8cfa5/pkg/server/status.go#L2073)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Can we possibly make a `struct` type to more clearly keep track of all this information? 

It seems like the key for all these maps is the `catalog.Descriptor` ID, so instead of having multiple maps, can we have a single struct type that encompasses all these attributes (db names, table names, index names, and parent ID), and then just use a single map? Something like

```
type HotRangeReportMeta struct {
    dbName string
    tableName string
    indexNames map[uint32]string
    parentId int32
}
rangeReportMetas := make(map[uint32]HotRangeReportMeta)
...
for _, desc := range descrs {
    rangeReportMetas[uint32(desc.GetId())] = HotRangeReportMeta{
        ...
    }
} 
```

Then the maintainer can get all the info they need with a single map access. Just an idea, let me know what you think. We could also potentially define functions on this type to encapsulate some of the more complex access patterns of this data, although maybe that's overkill :D 

</blockquote></details>

Done.
___
*[pkg/server/status.go, line 2106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mt4clwCDuSYv3Petrdg:-MtD3olA1092PRCHfCA-:b2nhll9) ([raw file](https://github.com/cockroachdb/cockroach/blob/896d8a409804e251fd6c7f73255df5f4f2d8cfa5/pkg/server/status.go#L2106)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Nothing stands out to me that would allow us to avoid this, I don't see it as too big of an issue, personally :) 
</blockquote></details>

Ok.


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
> One general question I wanted to pose here is have we considered how this might work in a serverless context? 

Just noting - we (myself, @koorosh, @rimadeodhar) had a chat about this today. The effort involved to make this work in a serverless context will begin with our work to get a minimal version of `debug.zip` working for serverless tenants. Since debug.zip uses the original HotRanges endpoint, which is also used by this new version, the work we do on the original endpoint should carry over capabilities to the new version as well.

The effort involved should be somewhat minimal - we just need to filter down the ranges/replicas reported on to only those with the appropriate tenant key prefix. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-:-MtEL1VDAd3lYl0w73Z_:b-ye3p1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dhartunian)



<!-- Sent from Reviewable.io -->

koorosh committed:
server, ui: use POST request method for HotRangesV2 service

Initially, `HotRangesV2` service in status server was defined
to use GET method to handle HTTP request. It was convenient way
to display response data in Advanced debugging page since it allowed
to render response body right on to page.

But now, hot ranges will be used on user facing page and request
dispatching for hot ranges api should follow generic workflow:
initialize `HotRangesRequest` protobuf message and dispatch request
with `src/util/api` service. This restriction forces to use POST
method (since GET method doesn't allow to provide request body).

In addition, Hot Ranges debugging page is refactored to use `api`
service.

Release note: None

Release justification: bug fixes and low-risk updates
to new functionality

abarganier reviewed (commented):
Nice work, I think this is almost ready. I just have a few docstring nits and an error condition that I think we should probably log. 

There are also some test failures in the CI build - have you tried running `make testrace` or `make stressrace` locally?  I've re-triggered them in case it was a spurious failure, but if not, let me know if I can help debug :thumbsup: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-:-Mtj22YVABjnsQgxPQgU:bw3uq0r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, and @koorosh)

___
*[pkg/server/status.go, line 2101 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj-o_50NfANuNMUyet:-Mtj-o_50NfANuNMUyeu:b-68wtjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/status.go#L2101)):*
> ```Go
> 				_, tableID, err := s.sqlServer.execCfg.Codec.DecodeTablePrefix(r.Desc.StartKey.AsRawKey())
> 				if err != nil {
> 					continue
> ```

nit: should we consider at least logging the error here? 



_Code quote:_
```Go
					continue
```
___
*[pkg/server/serverpb/status.proto, line 1071 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj0bKj9p0NzJnQ7a3D:-Mtj0bKj9p0NzJnQ7a3E:b-tawzty) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/serverpb/status.proto#L1071)):*
> ```protobuf
>     // API: PUBLIC ALPHA
>     double queries_per_second = 2;
>     // LeaseholderNodeID indicates on Node ID that contains replica that is leaseholder
> ```

nit: `// LeaseholderNodeID indicates the Node ID that is the current leaseholder for the given range.` 



_Code quote:_
```protobuf
// LeaseholderNodeID indicates on Node ID that contains replica that is leaseholder
```
___
*[pkg/server/serverpb/status.proto, line 1129 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj1FA49NGzktI1LOyR:-Mtj1FA49NGzktI1LOyS:bkiza58) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/serverpb/status.proto#L1129)):*
> ```protobuf
>   // HotRange message describes a single hot range, ie its QPS, node ID it belongs to, etc.
>   message HotRange {
>     // range_id indicates Range ID that's identified as hot range
> ```

nit: for this docstring and those below, let's make sure the sentences end with `.` periods to stay consistent with the rest of the file 
___
*[pkg/server/serverpb/status.proto, line 1134 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj0qiA2HZs7rQNMRpn:-Mtj0qiA2HZs7rQNMRpo:b-7glcjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/serverpb/status.proto#L1134)):*
> ```protobuf
>       (gogoproto.customname) = "RangeID"
>     ];
>     // node_id indicates on node that contains current hot range
> ```

nit: `// node_id indicates the node that contains the current hot range.`



_Code quote:_
```protobuf
// node_id indicates on node that contains current hot range
```
___
*[pkg/server/serverpb/status.proto, line 1144 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj10JB-L544MCShyJa:-Mtj10JB-L544MCShyJb:b-rsxcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/serverpb/status.proto#L1144)):*
> ```protobuf
>       (gogoproto.customname) = "QPS"
>     ];
>     // table_name indicates table which data is stored in this hot range
> ```

nit: `// table_name indicates the SQL table that the range belongs to.` 



_Code quote:_
```protobuf
// table_name indicates table which data is stored in this hot range
```
___
*[pkg/server/serverpb/status.proto, line 1155 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj1UV--859fFuXPl-V:-Mtj1UV--859fFuXPl-W:bov29oe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/serverpb/status.proto#L1155)):*
> ```protobuf
>         "github.com/cockroachdb/cockroach/pkg/roachpb.NodeID"
>     ];
>     // leaseholder_node_id indicates on Node ID that contains replica that is a leaseholder
> ```

nit: `// leaseholder_node_id indicates the Node ID that is the current leaseholder for the given range.` 



_Code quote:_
```protobuf
/ leaseholder_node_id indicates on Node ID that contains replica that is a leaseholder
```


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
Thanks for pointing out on this, I didn't run them. Will run locally and to check what's wrong

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-:-MtmepW4EVb24joM9LMk:bzdbqpm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dhartunian)

___
*[pkg/server/status.go, line 2101 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj-o_50NfANuNMUyet:-Mtme5GI5qs_qoCAJgoa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/status.go#L2101)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: should we consider at least logging the error here? 
</blockquote></details>

Done.
___
*[pkg/server/serverpb/status.proto, line 1071 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj0bKj9p0NzJnQ7a3D:-MtmehGoEOf0aF-XQgzM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/serverpb/status.proto#L1071)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: `// LeaseholderNodeID indicates the Node ID that is the current leaseholder for the given range.` 
</blockquote></details>

Done.
___
*[pkg/server/serverpb/status.proto, line 1129 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj1FA49NGzktI1LOyR:-MtmeMYC20QuEUYF3OC7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/serverpb/status.proto#L1129)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: for this docstring and those below, let's make sure the sentences end with `.` periods to stay consistent with the rest of the file 
</blockquote></details>

Done.
___
*[pkg/server/serverpb/status.proto, line 1134 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj0qiA2HZs7rQNMRpn:-MtmeQoK33I_fnjS6yoH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/serverpb/status.proto#L1134)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: `// node_id indicates the node that contains the current hot range.`
</blockquote></details>

Done.
___
*[pkg/server/serverpb/status.proto, line 1144 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj10JB-L544MCShyJa:-MtmeVYz5agAFJR5lSMU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/serverpb/status.proto#L1144)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: `// table_name indicates the SQL table that the range belongs to.` 
</blockquote></details>

Done.
___
*[pkg/server/serverpb/status.proto, line 1155 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mtj1UV--859fFuXPl-V:-MtmeaSJ281jityrdTb0:b-mlrxhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2149fcf4a2b4196b40f3006f9147e9f0490929b3/pkg/server/serverpb/status.proto#L1155)):*
> // leaseholder_node_id indicates the Node ID that is the current leaseholder for the given range.
Done.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
Can we squash some of the commits in this PR? We shouldn't have any "nit" commits in history. I'd prefer if we just made this PR a single commit but I can understand if you want to break it up into 2 or 3. Right now it's too many.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-:-MtsLlKL4a0iB05BASql:b-mlcnf8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dhartunian)



<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
Thanks David for helping polish this PR 🙏🏼. Agree, nit commits don't looks appropriate here. Also I'm wondered if `bors` squashes commits during merge? 🤔

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-:-MttBQ243VpJQ8dwOPdk:boa3bvf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dhartunian)



<!-- Sent from Reviewable.io -->

abarganier reviewed (approved):
Nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. Yes, I do think bors squashes during merge AFAIK. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-:-MuC69e36oVifi2S2L9i:b-y6dkmx)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier and @dhartunian)



<!-- Sent from Reviewable.io -->

dhartunian commented:
bors does not squash commits the way we have it configured. bors creates a merge commit with a description that contains all the commit descriptions of the branches it's merging concatenated together. your individual commits remain in the tree.

new set of commits look much nicer :) thx for amending.

knz commented:
Ok with just 128 ranges per store, and 1 response per store, the amount of data is more limited than I had feared. At least it's not significantly increasing with the number of ranges.

_however_,  in a large cluster there will be many stores. So I  don't think we should collect the data across all stores in a single response. PAgination will help with scanning all stores.

koorosh commented:
> Ok with just 128 ranges per store, and 1 response per store, the amount of data is more limited than I had feared. At least it's not significantly increasing with the number of ranges.
> 
> _however_, in a large cluster there will be many stores. So I don't think we should collect the data across all stores in a single response. PAgination will help with scanning all stores.

It is essential to collect hot ranges with highest QPS over all nodes, that's where pagination doesn't allow scan only particular nodes iteratively.

knz commented:
But we _do_ need to scan over all stores to aggregate the hot ranges, right? In any case?

Like, there's no one operation which atomically, in constant time, can find the top 128 ranges across all cluster stores. There's got to be a scan _somewhere_. That scan should be resistent to transient failures in the cluster.

knz commented:
In case it wasn't clear:

- from an API design perspective, I think the front-end should know the hot ranges _per store_, not aggregated across the whole cluster. That's because the SQL DBA needs to know the mapping of ranges _to regions_ (and thus use the mapping of store to regions), which needs the per-store breakdown.
- this means that in any case, the front-end needs to pull per-store hot ranges from the back-end
- and that pull will do so over all stores in the cluster. And that's where I say we need pagination. We shouldn't retrieve per-store data _across all stores_ in one response.

koorosh commented:
Thank you for making it clear! One last thing I want to clarify that in this case front-end will send multiple requests to backend for every node/store. For large clusters with dozens of nodes it might result in huge amount of consecutive or concurrent request per each _store_. 
Would it be acceptable to paginate per node and include store id per range in response?

knz commented:
> front-end will send multiple requests to backend for every node/store

Why? Can't we batch groups of results at a time?

> Would it be acceptable to paginate per node and include store id per range in response?

Yes! The choice of pagination key should be done differently depending on what you're paginating. If the unit is the store, you'll want to use the store ID as key (and perhaps node ID, I'm not sure).

koorosh commented:
Sorry, probably I misunderstood your message. Then current implementation does what you suggest, it requests all hot ranges in cluster (visits every node/store) and can be extended (as next iteration) to narrow down requests to specific nodes/stores only (this is not a part of V1 iteration for this API), and in this case it will send requests to specific nodes and visit particular stores.

knz commented:
but does it paginate?

koorosh commented:
It doesn't, I thought to use a simplified approach by parametrization of requests in the same way as pagination, but without ability to go back and forth page by page.

To follow the same pattern, I think following pagination can be sufficient to use: https://github.com/cockroachdb/cockroach/blob/a950be5e3399abfb02a05b665fd81b965cc58d11/pkg/server/status.go#L2215

koorosh committed:
server: add pagination to hot ranges API

This change extends hot ranges API to support pagination
of responses. It is also possible to avoid pagination when
page size is set to 0.

Release note: None

Release justification: bug fixes and low-risk updates
to new functionality

koorosh commented:
fyi, updated PR with additional change to implement pagination (last commit).

Santamaura reviewed (commented):
> Reviewed 1 of 3 files at r22, 1 of 5 files at r25, 1 of 8 files at r26, 12 of 12 files at r27, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @dhartunian, @koorosh, and @zachlite)

___
*[pkg/server/api_v2_ranges.go, line 437 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mw29wVLCY1VIgnsCP3N:-Mw29wVLCY1VIgnsCP3O:bd9h063) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc40744e4840f97c720069030a3093f518a1fb4/pkg/server/api_v2_ranges.go#L437)):*
> ```Go
> 	IndexName         string           `json:"index_name"`
> 	SchemaName        string           `json:"schema_name"`
> 	ReplicaNodeIDs    []roachpb.NodeID `json:"replica_node_ids"`
> ```

Were we still planning to include stores and/or locality?


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @dhartunian, @Santamaura, and @zachlite)

___
*[pkg/server/api_v2_ranges.go, line 437 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mw29wVLCY1VIgnsCP3N:-Mw2PyFN1hHZzB2m0978:b-zcoy9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc40744e4840f97c720069030a3093f518a1fb4/pkg/server/api_v2_ranges.go#L437)):*
<details><summary><i>Previously, Santamaura (Alex Santamaura) wrote…</i></summary><blockquote>

Were we still planning to include stores and/or locality?
</blockquote></details>

Correct, but I do prefer to add this change in separate PR. This one is already big enough :)


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
The overall structure is nice! 

I still have a question about the number of SQL objects and the wisdom of pulling all object descriptors at once.

(I'm also thinking about this from the perspective of a console user: how common is it to want to see hot ranges for all databases/tables? Are they not viewing this information in the context of a specific database, schema or table, which could be used as filter when retrieving the descriptors?)

> Reviewed 1 of 5 files at r25, 4 of 12 files at r27.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-:-Mw7lXvaBHuG2CE3SLhf:b-ujgalo)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @dhartunian, @koorosh, @Santamaura, and @zachlite)

___
*[pkg/server/status.go, line 2174 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mw7lHuP3aH5SAApntGT:-Mw7lHuP3aH5SAApntGU:b1ublxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc40744e4840f97c720069030a3093f518a1fb4/pkg/server/status.go#L2174)):*
> ```Go
> 		ctx, s.sqlServer.internalExecutor, s.db,
> 		func(ctx context.Context, txn *kv.Txn, descriptors *descs.Collection) error {
> 			all, err := descriptors.GetAllDescriptors(ctx, txn)
> ```

We probably want to find a way to paginate over the object descriptors too. If there's tens of thousands of them, that's going to be disruptive too.


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
@knz , thanks for your feedback! Most common users problem (described here: https://cockroachlabs.atlassian.net/browse/CRDB-2323) that hot ranges cause node restarts, so I think it is important to provide node-wide information about hot ranges to narrow down to specific DB/table/index.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-:-MwB6IzbCcKrtBQdqVgP:b-ratww6)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @dhartunian, @koorosh, @Santamaura, and @zachlite)



<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @dhartunian, @knz, @Santamaura, and @zachlite)

___
*[pkg/server/status.go, line 2174 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mw7lHuP3aH5SAApntGT:-MwW3L-G7PcOfEkxzYhL:bze8m4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc40744e4840f97c720069030a3093f518a1fb4/pkg/server/status.go#L2174)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

We probably want to find a way to paginate over the object descriptors too. If there's tens of thousands of them, that's going to be disruptive too.
</blockquote></details>

@knz , is it an option to consider pagination over object descriptors as a follow up task and implement it in scope of separate issue? 
I'd like to get current implementation ready and then incrementally improve it.  


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @AlexTalks, @dhartunian, @knz, @Santamaura, @tbg, and @zachlite)

___
*[pkg/server/status.go, line 2174 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mw7lHuP3aH5SAApntGT:-MwXYt7hEpDNGcSmIy4t:b-hgxkzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc40744e4840f97c720069030a3093f518a1fb4/pkg/server/status.go#L2174)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

@knz , is it an option to consider pagination over object descriptors as a follow up task and implement it in scope of separate issue? 
I'd like to get current implementation ready and then incrementally improve it.  
</blockquote></details>

@tbg, @alextalks , (since Raphael is out), could you please advise if we should keep using `getAllDescriptors` in this API or should we consider this additional improvement as optional for now and implement it later? Appreciate any help on this since I'm not confident enough to make my own decision.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> Reviewed 3 of 12 files at r27, 1 of 4 files at r28.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @AlexTalks, @dhartunian, @knz, @koorosh, @Santamaura, @tbg, and @zachlite)

___
*[pkg/server/api_v2_ranges.go, line 505 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mwc6ppC5h30TPOOhCxz:-Mwc6ppC5h30TPOOhCy-:bnl42v6) ([raw file](https://github.com/cockroachdb/cockroach/blob/743219a35cd76bfb90e725773d3253de4e796570/pkg/server/api_v2_ranges.go#L505)):*
> ```Go
> 		}
> 
> 		var hotRangeInfos []hotRangeInfo
> ```

You can pre-allocate this with `make([]hotRangeInfo, len(resp.Ranges))` and then insert into it with the `i` you get from `range`. That keeps you from growing/copying the array a bunch via `append`.
___
*[pkg/server/serverpb/status.proto, line 1876 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mwc7_nB6CH6jda6E9eR:-Mwc7_nB6CH6jda6E9eS:bgdck6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/743219a35cd76bfb90e725773d3253de4e796570/pkg/server/serverpb/status.proto#L1876)):*
> ```protobuf
>   }
> 
>   rpc HotRangesV2WithPagination(HotRangesRequest) returns (HotRangesResponseV2) {
> ```

This RPC is unimplemented...I assume you can remove it since the regular endpoint contains the pagination logic, right?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @AlexTalks, @dhartunian, @knz, @koorosh, @Santamaura, and @zachlite)

___
*[pkg/server/status.go, line 2174 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mw7lHuP3aH5SAApntGT:-Mwf7B4cC9_45YrFePvW:b-ovohaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc40744e4840f97c720069030a3093f518a1fb4/pkg/server/status.go#L2174)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

@tbg, @alextalks , (since Raphael is out), could you please advise if we should keep using `getAllDescriptors` in this API or should we consider this additional improvement as optional for now and implement it later? Appreciate any help on this since I'm not confident enough to make my own decision.
</blockquote></details>

That seems ok to me as long as it's adequately tracked.


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @AlexTalks, @dhartunian, @knz, @Santamaura, and @zachlite)

___
*[pkg/server/api_v2_ranges.go, line 505 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mwc6ppC5h30TPOOhCxz:-Mx-DHT3ESb66mbZ3wmw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/743219a35cd76bfb90e725773d3253de4e796570/pkg/server/api_v2_ranges.go#L505)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

You can pre-allocate this with `make([]hotRangeInfo, len(resp.Ranges))` and then insert into it with the `i` you get from `range`. That keeps you from growing/copying the array a bunch via `append`.
</blockquote></details>

Done.
___
*[pkg/server/status.go, line 2174 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mw7lHuP3aH5SAApntGT:-Mx-DPRj7MZq-vMhvDfS:b-7eklps) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc40744e4840f97c720069030a3093f518a1fb4/pkg/server/status.go#L2174)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

That seems ok to me as long as it's adequately tracked.
</blockquote></details>

Ack, thank you!
___
*[pkg/server/serverpb/status.proto, line 1876 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mwc7_nB6CH6jda6E9eR:-Mx-DVW1DNpnCTlFVEOm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/743219a35cd76bfb90e725773d3253de4e796570/pkg/server/serverpb/status.proto#L1876)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This RPC is unimplemented...I assume you can remove it since the regular endpoint contains the pagination logic, right?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74377)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier, @AlexTalks, @dhartunian, @knz, @Santamaura, and @zachlite)

___
*[pkg/server/status.go, line 2174 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74377#-Mw7lHuP3aH5SAApntGT:-Mx-RG0r8wAKo5mTj-5f:b-l4ib1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc40744e4840f97c720069030a3093f518a1fb4/pkg/server/status.go#L2174)):*
<details><summary><i>Previously, koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

Ack, thank you!
</blockquote></details>

https://github.com/cockroachdb/cockroach/issues/77126


<!-- Sent from Reviewable.io -->

koorosh commented:
bors r+

knz commented:
LGTM

koorosh commented:
bors retry

ajwerner commented:
bors r-

This checks in a `.pb.gw.go` file which we usually don't check in and it leads to skew with  #76592

koorosh commented:
bors r+
